module/zfs/vdev_disk.c

```diff
@@ -85,50 +85,51 @@ vdev_bdev_mode(int smode)
 }
 #endif /* HAVE_OPEN_BDEV_EXCLUSIVE */
 
-/* The capacity (in bytes) of a bdev that is available to be used by a vdev */
+/*
+ * Returns the usable capacity (in bytes) for the partition or disk.
+ */
 static uint64_t
-bdev_capacity(struct block_device *bdev, boolean_t wholedisk)
+bdev_capacity(struct block_device *bdev)
 {
-	struct hd_struct *part = bdev->bd_part;
-	uint64_t sectors = get_capacity(bdev->bd_disk);
-	/* If there are no paritions, return the entire device capacity */
-	if (part == NULL)
-		return (sectors << SECTOR_BITS);
+	return (i_size_read(bdev->bd_inode));
+}
 
-	/*
-	 * If there are partitions, decide if we are using a `wholedisk`
-	 * layout (composed of part1 and part9) or just a single partition.
-	 */
-	if (wholedisk) {
-		/* Verify the expected device layout */
-		ASSERT3P(bdev, !=, bdev->bd_contains);
-		/*
-		 * Sectors used by the EFI partition (part9) as well as
-		 * partion alignment.
-		 */
-		uint64_t used = EFI_MIN_RESV_SIZE + NEW_START_BLOCK +
-		    PARTITION_END_ALIGNMENT;
-
-		/* Space available to the vdev, i.e. the size of part1 */
-		if (sectors <= used)
-			return (0);
-		uint64_t available = sectors - used;
-		return (available << SECTOR_BITS);
+/*
+ * Returns the maximum expansion capacity of the block device,  When the
+ * vdev has been created as a 'wholedisk' then expansion may be possible.
+ * Before any expansion is performed the partition layout is verified to
+ * confirm the original layout (-part1 and -part9).  If everything checks
+ * out the primary partition will be resized and the reserved partition
+ * relocated to the new end of device as part of 'zpool online -e'.
+ */
+static uint64_t
+bdev_max_capacity(struct block_device *bdev, uint64_t wholedisk)
+{
+	uint64_t psize;
+	int64_t available;
+
+	if (wholedisk && bdev->bd_part != NULL && bdev != bdev->bd_contains) {
+		available = i_size_read(bdev->bd_contains->bd_inode) -
+		    ((EFI_MIN_RESV_SIZE + NEW_START_BLOCK +
+		    PARTITION_END_ALIGNMENT) << SECTOR_BITS);
```

Comment:
My understanding when I worked on this was that `EFI_MIN_RESV_SIZE + NEW_START_BLOCK + PARTITION_END_ALIGNMENT` is a conservative estimate of the non-part0 space on the device. 

In `zpool_label_disk` the size of part0 is effectively computed as `P2ALIGN(total_size - (EFI_MIN_RESV_SIZE + NEW_START_BLOCK), PARTITION_END_ALIGNMENT)`. The size of the padding caused by the alignment can be anything from 0 to `PARTITION_END_ALIGNMENT`. By assuming we were padded with the full `PARTITION_END_ALIGNMENT` bytes, we actually report a size for part0 that could be slightly smaller than is really available. However, I don't think this is a big issue because this is just the size reported to the user as expandsize and once the pool is actually expanded, it will use as much space as fits evenly into metaslabs.

Reply:
Interesting - can you describe the circumstances of the failure a little more? Was `/dev/sddj` ever expanded? Is it still showing up with the latest version of this patch?

I'm not sure why this calculation would ever end up determining the size of the partition. My understanding is that `efi_use_whole_disk` is where the partition size is updated after an expand and that doesn't use what we computed for `asize`.

Reply:
Quite simply, cannot create a pool on a 10TB disk using whole-disk method. Prior to the commit for expansion, we were able to create such a pool. It is unclear to me whether this is noticeable on a non-debug kernel.

Reply:
@richardelling can you post the exact size of the device in bytes, I'd like to see if I can reproduce what your seeing.

Reply:
@richardelling I believe I see what's going on, and this commit should in fact resolve the issue accidentally introduced by the first expansion commit.  It was possible `bdev_capacity()` could slightly over report the partition size leading to the ASSERT your seeing in the `wholedisk` case.  This is no longer possible since `bdev_capacity()` was split in to `bdev_capacity()` and `bdev_max_capacity()`.